### DISCUSSION TITLE
Pixi massive data drawing, how to share geometry?

### DISCUSSION THREAD

**Discussion by pcbren1987** [pixijs/pixijs#8358] [Q&A]
There are two schemes as follows:
With scheme B, the FPS is improved, but the memory consumption is very large. Can we use scheme B and share shape scheme at the same time?

A. 50K Graphic With shared geometry, FPS is only 18
![image](https://user-images.githubusercontent.com/89827437/169748866-6ab4d92e-7891-485c-9c35-e86d9f4c81b4.png)

B. Only one Pixi is instantiated Graphics, add 50K circle, FPS can reach 60
![image](https://user-images.githubusercontent.com/89827437/169747661-a910f054-3892-478a-8c6a-0a72bae868af.png)

**Comment by miltoncandelero** [comment]
For some extra FPS try setting the parent container `interactiveChildren` to false.
The slowdown is probably tied to the iteration and transform of the many displayobjects 🤔

**Reply by pcbren1987** [reply]
FPS still does not improve after interactiveChildren is set  is there any other way
![image](https://user-images.githubusercontent.com/89827437/169842767-1bcd1df2-7011-48be-a85b-d17f79b01c75.png)

**Comment by bigtimebuddy** [comment]
Pixi has an [adaptive curve triangulation](https://jsfiddle.net/bigtimebuddy/8f6cLbrq/), which means that for curved objects like circles triangulation is based on the relative size of the radius. Why is this relevant? In your first example, you are rendering a 1px radius circle. Likely sharing this geometry isn't really a significant savings. It's probably not that much more complicated than a 2-triangle Sprite. 

Obviously there are trade-off for anything rendering. This is basically what you found:

1. Scheme A - lower fps, but objects can be moved around independently. Objects are easily culled off screen (which will improve performance)
2. Scheme B - high memory, very poor performance on redraws, unable to cull off screen objects

If have a bunch of very tiny identical, static objects, you should be using a ParticleContainer. That will give you the best performance per number of objects. When you start to get into the dozens of thousands of objects, all the specifics matter a lot because of how you optimize.

**Reply by pcbren1987** [reply]
In order to test whether there is the best combination of the two drawing schemes of  50k.

In fact, we use Pixi to draw 2D PCB graphics (similar to CAD, the graphics are vector graphics, which can be enlarged, reduced and translated). At present, it is used for PCB drawing, and scheme B is adopted, but the existing problem is that the memory occupation is huge. We haven't found the best solution yet,

The following figure is a 2D PCB drawing drawn with Pixi, which can reach 300K elements at most,
![image](https://user-images.githubusercontent.com/89827437/169959659-bd902b3b-11bd-44df-a8dc-a293207ad471.png)

Does ParticleContainer support Graphic drawing? There are a lot of round or round hats, and no better solution has been found.

![image](https://user-images.githubusercontent.com/89827437/169963570-05ba83e6-eecf-4c86-934d-5f3a4c0e0c8d.png)

**Comment by andrewstart** [comment]
Your setting of `interactiveChildren` is actually incorrect - the parent of all of the circle instances needs to have it set, `app.stage` in your example.
As far as combining a single graphic with squared off lines with shared instance circles, I don't see why that wouldn't work - you'd just be responsible for drawing all of your own round caps. You might find that using semicircles instead gets a little more framerate due to lower overdraw (but I'm not sure about that).
